### PR TITLE
abstract the string slice struct to stringutils package

### PR DESCRIPTION
**Pull request by mountkin**
The logic of `runconfig.Entrypoint` and `runconfig.Command` are almost the same.
While I'm working on #15780 I find that I also need the string or slice parsing utils.
So I think abstracting the parsing code to a business independent package may be useful for other similar functions.

### CODE REVIEW

**Comment by calavera** [comment]
Very nice. LGTM

**Comment by aaronlehmann** [comment]
I like the idea of combining these types. But I'm not sure I see the value of having an abstraction around `[]string` to replace them. I'd think the original idea of having the specific types `Entrypoint` and `Command` was to abstract away the fact that they are implemented as `[]string`, and let them evolve in the future (add new fields, etc). But if we change the type to `StrSlice`, then they are defined as being `[]string`, but in a roundabout way.

What about just using `[]string` as the type, and having the `NewStrSlice` convenience function return `[]string` directly? Then we wouldn't need the extra wrappers for JSON marshalling/unmarshalling, `Len`, `Slice`, etc.

**Comment by mountkin** [comment]
@aaronlehmann we can't use []string directly because the cmd or entrypoint may be a string or a slice of strings. If we use []string a BC break will be introduced.

If in the future we need to add more fields to the structure, we can just create a new struct in runconfig and embed the StrSlice struct.

**Comment by calavera** [comment]
@aaronlehmann, @mountkin is right. We need this struct to support arguments that can be a string and a slice of strings at the same time.

File: daemon/daemon.go [hunk 1]
```diff
@@ -437,7 +438,7 @@ func (daemon *Daemon) generateHostname(id string, config *runconfig.Config) {
 	}
 }
 
-func (daemon *Daemon) getEntrypointAndArgs(configEntrypoint *runconfig.Entrypoint, configCmd *runconfig.Command) (string, []string) {
+func (daemon *Daemon) getEntrypointAndArgs(configEntrypoint *stringutils.StrSlice, configCmd *stringutils.StrSlice) (string, []string) {
```
Comment (cpuguy83) [inline daemon/daemon.go]: nit, only need to define the type here once since both args are the same type.

**Comment by runcom** [comment]
This may also be useful if we should continue supporting old []string which accept a string as well (see #14338)

LGTM

**Pull request merged**